### PR TITLE
Add Symbol and timedelta interop features to cirq.Duration

### PR DESCRIPTION
**Pull request by Strilanc**
- Introduce `DURATION_LIKE` type to replace many instances of `Union[Duration, timedelta]`
- Add micros and millis as supported units
- Improve the `__str__` and `__repr__` methods to use better scaled units
- Allow `_picos` to be symbolic, add `_parameterized_` and `_resolve_parameters_` methods
- Allow comparisons with timedelta and with 0
- Deprecate `Duration.create` in favor of Duration's constructor
- Add `__bool__` method for easy non-zero checking

```
 >>> print(cirq.Duration(micros=1.5 * sympy.Symbol('t')))
(1500.0*t) ns
```

Fixes https://github.com/quantumlib/Cirq/issues/2279

### CODE REVIEW

File: cirq/value/duration.py [hunk 1]
```diff
@@ -13,120 +13,234 @@
...
-            return duration
-        if isinstance(duration, timedelta):
-            duration_in_picos = duration.total_seconds() * 10**12
-            return cls(picos=duration_in_picos)
+        return Duration(duration)
 
-        raise TypeError(
-            'Only datetime.timedelta and cirq.Duration are supported.')
+    def _is_parameterized_(self):
+        from cirq import protocols
```
Comment (MichaelBroughton) [inline cirq/value/duration.py]: Why is this import happening here ?

File: cirq/value/duration.py [hunk 2]
```diff
@@ -13,120 +13,234 @@
...
+        return protocols.is_parameterized(self._picos)
 
-    def total_picos(self) -> float:
+    def _resolve_parameters_(self, param_resolver):
+        from cirq import protocols
```
Comment (MichaelBroughton) [inline cirq/value/duration.py]: same as above

File: cirq/value/duration.py [hunk 3]
```diff
@@ -13,120 +13,234 @@
...
         return self.__mul__(other)
 
     def __truediv__(self, other) -> Union['Duration', float]:
-        if isinstance(other, (int, float)):
+        if isinstance(other, (int, float, sympy.Basic)):
             return Duration(picos=self._picos / other)
-        if isinstance(other, type(self)):
-            return self._picos / other._picos
+
+        other_duration = _attempt_duration_like_to_duration(other)
```
Comment (MichaelBroughton) [inline cirq/value/duration.py]: If div can produce a Duration or a Float should mul do the same ?

File: cirq/value/duration.py (hunk 1 above)
Comment (Strilanc) [reply cirq/value/duration.py]: I thought there'd be a circular reference, but upon testing there isn't. Removed.

File: cirq/value/duration.py (hunk 2 above)
Comment (Strilanc) [reply cirq/value/duration.py]: I thought there'd be a circular reference, but upon testing there isn't. Removed.

File: cirq/value/duration.py (hunk 3 above)
Comment (Strilanc) [reply cirq/value/duration.py]: Nope. second/second is a scalar, but second*second is neither a scalar nor a duration.

**Comment by Strilanc** [comment]
I added the DURATION_LIKE to _GateOperation.

**Pull request merged**